File: bindings/python/python/matter_vault/calls.py
"""Ready-to-submit on-chain call arguments (substrate-interface ``call_params``).

The builders take an ``Aad`` tag so a secret can't be stored under a different AAD
than it was sealed with.

These are for callers who submit with their own Substrate client.
:class:`~matter_vault.client.MatterClient` can submit them for you, which is the
shorter path for most callers.
"""

from typing import Tuple, Union

from .aad import Aad, aad_bytes
from .hexutil import to_hex

# (binding_id, capsule, proof, ct) — exactly what ``encrypt`` returns.
EncryptedSecret = Tuple[bytes, bytes, bytes, bytes]


def _payload(env: EncryptedSecret) -> dict:
    binding_id, capsule, proof, ct = env
    return {
        "binding_id": to_hex(binding_id),
        "capsule": to_hex(capsule),
        "proof": to_hex(proof),
        "ct": to_hex(ct),
    }


def store_secret(
    payload: EncryptedSecret, epoch: int, label: Union[str, bytes], aad: Union[Aad, str, bytes]
) -> dict:
    """``call_params`` for ``secrets.store_secret(payload, epoch, label, aad)``."""
    return {
        "payload": _payload(payload),
        "epoch": int(epoch),
        "label": to_hex(label.encode() if isinstance(label, str) else bytes(label)),
        "aad": to_hex(aad_bytes(aad)),
    }


def rotate_secret(
    secret_id: int, payload: EncryptedSecret, epoch: int, aad: Union[Aad, str, bytes]
) -> dict:
    """``call_params`` for ``secrets.rotate_secret(secret_id, payload, epoch, aad)``."""
    return {
        "secret_id": int(secret_id),
        "payload": _payload(payload),
        "epoch": int(epoch),
        "aad": to_hex(aad_bytes(aad)),
    }


ACCOUNT_ID_BYTES = 32


def grant_to_user(account: bytes) -> dict:
    """A ``GrantTarget`` naming an account."""
    if len(account) != ACCOUNT_ID_BYTES:
        raise ValueError(f"account must be {ACCOUNT_ID_BYTES} bytes, got {len(account)}")
    return {"User": to_hex(account)}


def grant_to_deployment(deployment: int) -> dict:
    """A ``GrantTarget`` naming a deployment.

    Authorizes whichever resource is currently assigned to it, so the owner need
    not name an account that only exists after assignment.
    """
    return {"Deployment": int(deployment)}


def grant_access(secret_id: int, target: dict) -> dict:
    """``call_params`` for ``secrets.grant_access(secret_id, target)``.

    ``target`` is a ``GrantTarget`` from :func:`grant_to_user` or
    :func:`grant_to_deployment`.

    The chain's ``GrantTarget<AccountId>`` is an **enum**, not a bare account id.
    This builder previously emitted a raw 32-byte ``grantee`` with a docstring
    claiming the chain would wrap it — nothing did, so the call data was dead on
    arrival. Nothing caught it because no end-to-end test exercised ``grant``.
    """
    return {"secret_id": int(secret_id), "target": target}


def revoke_access(secret_id: int, target: dict) -> dict:
    """``call_params`` for ``secrets.revoke_access(secret_id, target)``.

    The target must match the grant exactly, or this is a no-op on chain.
    Revocation is how a leaked signer is contained, so it is a first-class call
    rather than something callers hand-roll.
    """
    return {"secret_id": int(secret_id), "target": target}


def delete_secret(secret_id: int) -> dict:
    """``call_params`` for ``secrets.delete_secret(secret_id)``.

    Removes the secret and every grant on it. Owner only, and irreversible.
    """
    return {"secret_id": int(secret_id)}

File: packages/go/mattervault/calls.go
package mattervault

// Ready-to-submit on-chain call arguments for pallet-secrets.
//
// Go previously had no call builders at all — StoreSecret assembled its arguments
// inline, and the other four pallet-secrets calls were unreachable. These give the
// remaining calls a home, and give GrantTarget one shape shared with Rust,
// TypeScript, and Python.

import (
	"fmt"

	"github.com/centrifuge/go-substrate-rpc-client/v4/types"
)

// GrantTargetKind distinguishes the runtime's GrantTarget variants.
type GrantTargetKind uint8

const (
	// GrantUser names another user or a resource node by account.
	GrantUser GrantTargetKind = 0
	// GrantDeployment names a deployment, authorizing whichever resource is
	// currently assigned to it — so the owner need not name an account that only
	// exists after assignment.
	GrantDeployment GrantTargetKind = 1
)

// GrantTarget is who a secret is granted to.
//
// The chain's GrantTarget<AccountId> is an enum, not a bare account id. Earlier
// call builders in the other bindings emitted a raw 32-byte grantee, which the
// runtime cannot decode — the call data was dead on arrival, and nothing caught it
// because no end-to-end test exercised grant.
type GrantTarget struct {
	Kind GrantTargetKind
	// Account is set when Kind is GrantUser.
	Account []byte
	// Deployment is set when Kind is GrantDeployment.
	Deployment SecretID
}

// UserTarget builds a grant target naming an account.
func UserTarget(account []byte) (GrantTarget, error) {
	if len(account) != accountIDBytes {
		return GrantTarget{}, fmt.Errorf("account must be %d bytes, got %d", accountIDBytes, len(account))
	}
	return GrantTarget{Kind: GrantUser, Account: account}, nil
}

// DeploymentTarget builds a grant target naming a deployment.
func DeploymentTarget(deployment SecretID) GrantTarget {
	return GrantTarget{Kind: GrantDeployment, Deployment: deployment}
}

// String renders the target for logs and errors.
func (t GrantTarget) String() string {
	if t.Kind == GrantUser {
		return fmt.Sprintf("User(%s)", toHex(t.Account))
	}
	return fmt.Sprintf("Deployment(%s)", t.Deployment)
}

// encode renders the target as the SCALE enum the runtime decodes: a one-byte
// variant index followed by the variant's payload.
func (t GrantTarget) encode() ([]byte, error) {
	switch t.Kind {
	case GrantUser:
		if len(t.Account) != accountIDBytes {
			return nil, fmt.Errorf("grant target account must be %d bytes", accountIDBytes)
		}
		return append([]byte{byte(GrantUser)}, t.Account...), nil
	case GrantDeployment:
		le := t.Deployment.LEBytes()
		return append([]byte{byte(GrantDeployment)}, le[:]...), nil
	default:
		return nil, fmt.Errorf("unknown grant target kind %d", t.Kind)
	}
}

// secretsPayload is the runtime's EncryptedSecret composite.
type secretsPayload struct {
	BindingID types.Bytes
	Capsule   types.Bytes
	Proof     types.Bytes
	CT        types.Bytes
}

func newSecretsPayload(env EncryptedSecret) secretsPayload {
	return secretsPayload{
		BindingID: types.NewBytes(env.BindingID),
		Capsule:   types.NewBytes(env.Capsule),
		Proof:     types.NewBytes(env.Proof),
		CT:        types.NewBytes(env.CT),
	}
}

// StoreSecretCall builds `Secrets.store_secret(payload, epoch, label, aad)`.
func (c *ChainClient) StoreSecretCall(env EncryptedSecret, epoch uint32, label string, aad Aad) (types.Call, error) {
	return types.NewCall(
		c.meta,
		"Secrets.store_secret",
		newSecretsPayload(env),
		types.NewU32(epoch),
		types.NewBytes([]byte(label)),
		types.NewBytes(AadBytes(aad)),
	)
}

// RotateSecretCall builds `Secrets.rotate_secret(secret_id, payload, epoch, aad)`.
func (c *ChainClient) RotateSecretCall(id SecretID, env EncryptedSecret, epoch uint32, aad Aad) (types.Call, error) {
	return types.NewCall(
		c.meta,
		"Secrets.rotate_secret",
		types.NewU128(*id.BigInt()),
		newSecretsPayload(env),
		types.NewU32(epoch),
		types.NewBytes(AadBytes(aad)),
	)
}

// GrantAccessCall builds `Secrets.grant_access(secret_id, target)`.
func (c *ChainClient) GrantAccessCall(id SecretID, target GrantTarget) (types.Call, error) {
	encoded, err := target.encode()
	if err != nil {
		return types.Call{}, err
	}
	return types.NewCall(c.meta, "Secrets.grant_access", types.NewU128(*id.BigInt()), encoded)
}

// RevokeAccessCall builds `Secrets.revoke_access(secret_id, target)`.
//
// The target must match the grant exactly, or the call is a no-op on chain.
func (c *ChainClient) RevokeAccessCall(id SecretID, target GrantTarget) (types.Call, error) {
	encoded, err := target.encode()
	if err != nil {
		return types.Call{}, err
	}
	return types.NewCall(c.meta, "Secrets.revoke_access", types.NewU128(*id.BigInt()), encoded)
}

// DeleteSecretCall builds `Secrets.delete_secret(secret_id)`. Owner only, and
// irreversible.
func (c *ChainClient) DeleteSecretCall(id SecretID) (types.Call, error) {
	return types.NewCall(c.meta, "Secrets.delete_secret", types.NewU128(*id.BigInt()))
}
